Export function for ensuring docs are installed

chunned opened the pull request:
Originally discussed in #8.

Previously, checking `ensure_installed`  happened with statements included in `apidocs_open`. This breaks that logic out into a function that is then exposed to the user so they can call it when they like.

For instance, using Lazy package manager, setting `lazy = false` and calling
`require("apidocs").apidocs_ensure_install("ensure_installed"={"lua~5.4"}})` will download the Lua 5.4 docs on Nvim startup.

Default functionality is unchanged.

chunned commented:
> i think that ensure_install will in fact open the picker after running?

You're correct, and I think your callback idea is a good solution!

> and it would be nice to update the readme to mention this new feature.

Perhaps I'm overthinking this, but which of the two below would you prefer?

> - `require("apidocs").apidocs_ensure_install({ensure_installed={"lua~5.4"}})` - ensure the documentation for each provided language is installed (in this example, Lua 5.4 docs will be installed). You can pass in any number of languages to this function to ensure they are all present on your system.

or

> `require("apidocs").apidocs_ensure_install(opts)` - ensure the documentation for each provided language is installed. `opts` should be a table with the key `ensure_installed` and any number of values corresponding to languages to download, e.g. `{ensure_installed={"lua~5.4"}}`

Of course, I am open to any revisions of these two lines.

I likely won't have time to implement the callback function until next week, so if you would like to take over and implement it, that is all good with me. Otherwise, I can do it next time I have some spare time.

emmanueltouzery commented:
actually I would probably prefer it to work like that: `require..apidocs_ensure_install({"lua~5.4})`. So probably no need to write twice `ensure_installed`? that array can then be wrapped to be passed on later, internally.

the style between the two.. your call. maybe whatever is more similar to what's existing. if you really don't know.. maybe the first one? but it's really not a big concern for me.

chunned commented:
Managed to find a bit of time to complete these suggestions and all should now be good. Let me know if you'd like any other changes!

emmanueltouzery commented on lua/apidocs.lua:
```diff
@@ -47,26 +44,65 @@ local function apidocs_open(opts)
+  end
+
+  for _, source in ipairs(languages) do
+    if not vim.tbl_contains(installed_docs, source) then
```
this is going to install multiple sources concurrently, right? with the freezing of neovim already just installing one, i'm not super comfortable triggering multiple at once. the approach i had of recursively calling myself was maybe a little messy, but it ensured we only try installing one at once, i think that was better.

here we have this function triggering multiple async tasks (ultimately async because we invoke curl in an async manner), then getting back that "multithread" input back into pending_installations... i think what i had was simpler at runtime, i preferred that.

so... install one (no loop), and then call back ensure_install_and_then. if there is nothing to install, not even one, then call `cont`. That's how i would do it.

chunned commented:
> sorry for my slow feedback... unfortunately i have some more comments...

No worries - this is one of the first PR's I've ever opened, I'd rather get it exactly how you like!

These should be very easy changes, will try to integrate them by the end of the weekend.

chunned commented:
Fixed the return and modified `ensure_install_and_then`.

Regarding the local variables, they are still used, I've moved their definitions to be closer to where they are used so that's clearer. Note that it would be possible to break out the below snippet into its own function, `get_candidates` or something, so that `apidocs_open` is more readable:

```lua
function apidocs_open(opts)
  ...
  ...
  ...
  local docs_path = common.data_folder()
  local candidates = {}
  for _, name in ipairs(installed_docs) do
    local fs2 = vim.uv.fs_scandir(docs_path .. "/" .. name)
    while true do
      local name2, type2 = vim.uv.fs_scandir_next(fs2)
      if not name2 then
        break
      end
      if type2 == "file" and vim.endswith(name2, ".html.md") then
        local name_no_txt = name2:gsub("#.*$", "")
        table.insert(candidates, { display = name .. "/" .. name_no_txt, path = name .. "/" .. name2 })
      end
    end
  end
```

If you'd like me to change that, just let me know.

Thanks for your patience :smile:

emmanueltouzery commented:
we'll get there :crossed_fingers: 

i think an implementation of ensure_install_and_then like that would be simpler:

```lua
local function ensure_install_and_then(languages, slugs_to_mtimes, cont)
  local installed_docs = get_installed_docs()

  for _, source in ipairs(languages) do
    if not vim.tbl_contains(installed_docs, source) then
      if slugs_to_mtimes == nil then
        install.fetch_slugs_and_mtimes_and_then(function(slugs_to_mtimes)
          install.apidoc_install(source, slugs_to_mtimes, function()
            ensure_install_and_then(languages, slugs_to_mtimes, cont)
          end)
        end)
        return
      else
        install.apidoc_install(source, slugs_to_mtimes, function()
          ensure_install_and_then(languages, slugs_to_mtimes, cont)
        end)
        return
      end
    end
  end
  -- everything is installed, move on
  cont(slugs_to_mtimes)
end
```

it also makes sure slugs_to_mtimes is fetched only once, which is apparently broken with the current version :sob: 

of course, the first time ensure_install_and_then is called, slugs_to_mtimes should be passed in as nil.

compared to what you had to start with, get_installed_docs() is called multiple times, but it's only reading a few folders from the local disk.

chunned commented:
> it also makes sure slugs_to_mtimes is fetched only once, which is apparently broken with the current version 😭

Broken or in main or broken by my changes? I inspected the table in both and don't see anything wrong, but obviously you know the plugin better than myself

emmanueltouzery commented:
it was already broken before your changes, as far as i can tell, because slugs_to_mtimes is fetched multiple times, before your changes.

emmanueltouzery commented:
thank you for all the efforts and patience! i'll merge this now but immediately make a little followup commits with some nitpicks. thank you again!
